RecoHGCal/TICL/plugins/ClusterFilterByAlgoOrSize.h

```diff
@@ -0,0 +1,42 @@
+// Author: Marco Rovere - marco.rovere@cern.ch
+// Date: 11/2018
+// Copyright CERN
+
+#ifndef RecoHGCal_TICL_ClusterFilterByAlgoOrSize_H__
+#define RecoHGCal_TICL_ClusterFilterByAlgoOrSize_H__
+
+#include "DataFormats/CaloRecHit/interface/CaloCluster.h"
+#include "ClusterFilterBase.h"
+
+#include <memory>
+#include <utility>
+
+// Filter clusters that belong to a specific algorithm
+namespace ticl{
+class ClusterFilterByAlgoOrSize final : public ClusterFilterBase {
+ public:
+  ClusterFilterByAlgoOrSize(const edm::ParameterSet& ps)
+      : ClusterFilterBase(ps),
+        algo_number_(ps.getParameter<int>("algo_number")),
+        max_cluster_size_(ps.getParameter<int>("max_cluster_size")) {}
+  ~ClusterFilterByAlgoOrSize() override {};
+
+  std::unique_ptr<std::vector<std::pair<unsigned int, float>>> filter(
+      const std::vector<reco::CaloCluster>& layerClusters,
+      const std::vector<std::pair<unsigned int, float>>& availableLayerClusters) const override {
+    auto filteredLayerClusters = std::make_unique<std::vector<std::pair<unsigned int, float>>>();
+    for (auto const& cl : availableLayerClusters) {
+      if (layerClusters[cl.first].algo() == algo_number_ ||
+          layerClusters[cl.first].hitsAndFractions().size() <= max_cluster_size_)
```

Comment:
Ciao @slava77 ,
I'm not sure I understand your suggestion: do you mean moving the loop in the base-class and call an ad-hoc filter on each CaloCluster?

Reply:
yes, because the loop and related setup code has to be repeated in all derived variants of the ClusterFilterBase.

Reply:
yes, indeed, but maybe this loop w/o virtuals will vectorize?

Reply:
I'm not sure what is implied by lack of virtuals.
Simple function calls, even via a vtable can be vectorizable (or is it impossible?); access to layerClusters via a generally random list of indices does not look vectorizable by itself.

Reply:
@slava77 I'm not sure if there can be vectorization with vtable in between.
In any case, forcing the loop in the base will fix the logic for all derived (i.e. no comparison between 2 elements will ever be possible). I'm not sure this will ever be required, but I think this is good enough for the release.